librz/asm/p/asm_sh.c

```diff
@@ -1,78 +1,51 @@
 // SPDX-FileCopyrightText: 2010-2011 eloi <limited-entropy.com>
 // SPDX-License-Identifier: LGPL-3.0-only
+// SPDX-FileCopyrightText: 2022 Dhruv Maroo <dhruvmaru007@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
 
-#include <stdio.h>
-#include <stdarg.h>
-#include <string.h>
-#include <rz_types.h>
 #include <rz_lib.h>
 #include <rz_util.h>
 #include <rz_asm.h>
-#include "disas-asm.h"
-
-static unsigned long Offset = 0;
-static RzStrBuf *buf_global = NULL;
-static unsigned char bytes[4];
-
-static int sh_buffer_read_memory(bfd_vma memaddr, bfd_byte *myaddr, unsigned int length, struct disassemble_info *info) {
-	// this is obviously wrong. but how can we read arbitrary data @ memaddr from here?
-	memcpy(myaddr, bytes, length);
-	return 0;
-}
-
-int print_insn_shl(bfd_vma memaddr, struct disassemble_info *info);
-int print_insn_shb(bfd_vma memaddr, struct disassemble_info *info);
+#include "../arch/sh/disassembler.h"
+#include "../arch/sh/assembler.h"
 
-static int symbol_at_address(bfd_vma addr, struct disassemble_info *info) {
-	return 0;
-}
-
-static void memory_error_func(int status, bfd_vma memaddr, struct disassemble_info *info) {
-	//--
+static int disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
+	SHOp *dis_op = sh_disassembler(rz_read_ble16(buf, a->big_endian));
+	op->size = 2;
+	if (!dis_op) {
+		rz_strbuf_set(&op->buf_asm, "invalid");
+	} else {
+		char *disasm = sh_op_to_str(dis_op, a->pc);
+		rz_strbuf_set(&op->buf_asm, disasm);
+		free(disasm);
+	}
+	RZ_FREE(dis_op);
```

Comment:
`RZ_FREE()` also sets `dis_op` to `NULL`, since it's a local variable and isn't used anymore, just `free()` would be sufficient

Reply:
Yes, I just wanted to avoid any sort of use-after-free stuff (by mistake), so just to be sure.